doc/reference/trd-appid.md

```diff
@@ -0,0 +1,651 @@
+single Tock system.  Global application identifiers MUST persist
+across process restarts or reloads.
+
+If the verifier policy assigns the same application identifier to
+multiple process binaries, then the Tock kernel MUST NOT run more than
+one of them at any given time. Following the above example, the Tock
+kernel can run v1.1 or v1.2 of the application, but will not run both
+simultaneously. A kernel MAY perform this check by comparing Short IDs
+generated from application identifiers using the `Compress` trait
+(described below). Kernels using Short IDs to test collisions between
+application identifiers SHOULD implement `Compress` in a manner that
+minimizes cases when two different valid application identifiers
+compress to the same Short ID (e.g., taking the low-order bits
+of a strong cryptographic hash function).
+
+In cases when a process does not have any application
+credentials, the verifier policy MAY assign it a global or local
+application identifier. If the verifier policy does not assign a
+process binary an application identifier then the kernel MUST NOT load
+or run that process.
+
+Consider these five use cases.
+
+  1. A process binary that has no application credentials: it only
+  runs on kernels that are willing to load applications without
+  credentials (e.g., research systems). The verifier policy defines
+  that process binaries with no credentials have a local application 
+  identifier of a SHA256 hash of the application binary. 
+
+  1. The verifier policy defines that the global application
+  identifier of a process is the public key used to sign its process
+  binary. Application credentials for a process binary consist of a
+  TBF header containing this public key as well as a signature for the
+  process binary signed with the associated private key. Before
+  verifiying a signature in a TBF header, the verifier policy decides
+  whether to it accepts the associated public key. The verifier policy
+  assigns a global application identifier as the public key in the TBF
+  header.
+
+  1. Multiple separate process binaries that run concurrently need to
+  be signed with a single public key. Each process binary is
+  identified by a unique identifier I. Application credentials for
+  these process binaries consist of a TBF header containing an ECDSA
+  public key, the unique identifier I, and a signature over the
+  process binary and I signed with the private key corresponding to
+  the public key in the header. The kernel decides whether to accept a
+  particular public key for verification. The verifier policy assigns
+  a global application identifier as the concatenation of the public
+  key and I.
+
+  1. A Tock system wants to load the same process binary in 
+  two different processes at the same time. It cannot. Every process
+  binary has a single application identifier, and Tock will not run
+  two processes with the same application identifier.
```

Comment:
Are there details on which one Tock will run though?

Reply:
For records: Alistair and I talked about this in the OT call. The thought is to add a version number TBF header. If a process binary doesn't have a version number header, its number is 0. Processes are loaded in order of descending version numbers (highest first), with ties being resolved by which binary is at a lower address in flash.